File: server/Models/Product.cs
using server.Models.Enums;

namespace server.Models;

public class Product
{
    public Guid Id{get;set;}
    public string Name{get;set;}=string.Empty;
    public string? Brand{get;set;}
    public ProductType Type{get;set;}
    public ProductVisibility Visibility{get;set;} = ProductVisibility.Private;
    public ProductSourceType SourceType{get;set;}
    public ProductReviewStatus ReviewStatus{get;set;}=ProductReviewStatus.NotSubmitted;
    
    public Guid CategoryId{get;set;}
    public ProductCategory Category{get;set;} = null!;

    public Guid? OwnerId{get;set;}
    public User? Owner{get;set;}

    public double Calories{get;set;}
    public double Protein{get;set;}
    public double Fat{get;set;}
    public double Carbs{get;set;}
    public double? Sugar{get;set;}
    public double? Fiber{get;set;}
    public double? IronMg{get;set;}

    public ProductUnit DefaultUnit{get;set;}

    public double NutritionAmount{get;set;}
    public ProductUnit NutritionUnit{get;set;}

    public double? ServingSize{get;set;}
    public string? ServingDescription{get;set;}

    public DateTime CreatedAt{get;set;} = DateTime.UtcNow;
    public DateTime UpdatedAt{get;set;} = DateTime.UtcNow;

    public List<ProductStock> ProductStocks{get;set;}=new();
    public List<FoodEntry> FoodEntries{get;set;}=new();
    public List<DrinkEntry> DrinkEntries{get;set;}=new();

}

File: client/src/shared/types/product.ts
export type ProductType = "Food"  | "Drink";

export type ProductVisibility = "Public" | "Private";

export type ProductUnit = "Gram" | "Ml" | "Portion";

export type ProductSourceType = "Custom" | "Branded";

export type ProductReviewStatus = "NotSubmitted" | "PendingReview" | "Approved" | "Rejected";

export type ProductResponse = {
    id: string;
    name: string;
    brand: string | null;

    sourceType:ProductSourceType;
    reviewStatus: ProductReviewStatus;
    type: ProductType;
    visibility: ProductVisibility;

    categoryId: string;
    categoryName: string;

    ownerId: string | null;

    calories: number;
    protein: number;
    fat: number;
    carbs: number;

    sugar: number | null;
    fiber: number | null;
    ironMg: number | null;

    defaultUnit: ProductUnit;

    nutritionAmount: number;
    nutritionUnit: ProductUnit;

    servingSize: number | null;
    servingSizeUnit: ProductUnit | null;
    servingDescription: string | null;

    createdAt: string;
    updatedAt: string;
}

export type CreateProductRequest ={
    name: string;
    type: ProductType;
    brand?: string | null;
    categoryId: string;
    sourceType:ProductSourceType;

    calories: number;
    protein: number;
    fat: number;
    carbs: number;

    sugar?: number | null;
    fiber?: number | null;
    ironMg?: number | null;

    defaultUnit: ProductUnit;

    nutritionUnit: ProductUnit;
    nutritionAmount: number;

    servingSize?: number | null;
    servingDescription?: string | null;
}

export type UpdateProductRequest = CreateProductRequest;